### PR TITLE
[FLINK-19154] ApplicationDispatcherBootstrap clean up HA data only when necessary

### PR DESCRIPTION
**Pull request by kl0u**
## What is the purpose of the change

Currently in Application Mode when we shutdown the cluster we always clean up the HA data of a job. This also includes the cases when the job fails due to a transient error. This behaviour leads to the job not being able to restart even if HA is activated. This PR fixes the issue by only cleaning up the HA data when the job has explicitly reported reaching a terminal state (`FAILED`, `SUCCEEDED`, or `CANCELLED`), and not as a reaction to framework failures or other issues.

**NOTE TO REVIEWERS:** Maybe in some cases, we would like to not even shut down the cluster. But I do not think we can find all these cases and filter by specific exceptions so at least now we can go through the restart process without problems.

## Brief change log

The changes are split in 3 commits. The main part related to the problem is in the `ApplicationDispatcherBootstrap.runApplicationAndShutdownClusterAsync()` but there is also an important part included in the 3rd commit where the `ApplicationDispatcherBootstrap` only takes the `DispatcherGateway`.

## Verifying this change

This change is already covered by existing tests, such as the tests in the `ApplicationDispatcherBootstrapTest` that were modified.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

### CODE REVIEW

**Comment by kl0u** [comment]
@flinkbot run azure

File: flink-clients/src/main/java/org/apache/flink/client/deployment/application/exceptions/ApplicationFailureException.java [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+
+import static org.apache.flink.util.Preconditions.checkNotNull;
+import static org.apache.flink.util.Preconditions.checkState;
+
+/**
+ * Generic exception to signal the failure of an application.
+ * This is to be subclassed by more specific exceptions.
+ */
+@Internal
+public class ApplicationFailureException extends FlinkException {
```
Comment (tillrohrmann) [inline flink-clients/src/main/java/org/apache/flink/client/deployment/application/exceptions/ApplicationFailureException.java:38]: Instead of having multiple subclasses of `ApplicationFailureException` we could have a single `ApplicationExecutionException` which also contains the application status as a field. Then one could do the matching based on the `ApplicationStatus` instead of the different sub classes. This might be a bit more robust.

File: flink-clients/src/main/java/org/apache/flink/client/deployment/application/ApplicationDispatcherBootstrap.java [hunk 2]
```diff
@@ -155,8 +160,9 @@ public void stop() {
 						}
 
 						LOG.warn("Exiting with Application Status UNKNOWN: ", t);
-						return dispatcher.shutDownCluster(ApplicationStatus.UNKNOWN);
 
+						this.errorHandler.onFatalError(t);
```
Comment (tillrohrmann) [inline flink-clients/src/main/java/org/apache/flink/client/deployment/application/ApplicationDispatcherBootstrap.java:161]: Maybe add an exception message `new FlinkException("Application failed unexpectedly.", t)` or so.

File: flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/Dispatcher.java [hunk 3]
```diff
@@ -208,7 +210,8 @@ public void onStart() throws Exception {
 			throw exception;
 		}
 
-		dispatcherBootstrap.initialize(this, this.getRpcService().getScheduledExecutor());
+		this.dispatcherBootstrap = this.dispatcherBootstrapFactory.apply(shutDownFuture::completeExceptionally);
+		this.dispatcherBootstrap.initialize(this, this.getRpcService().getScheduledExecutor());
```
Comment (tillrohrmann) [inline flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/Dispatcher.java]: Could `initialize` be merged with the creation step of the `ApplicationDispatcherBootstrap`?

File: flink-clients/src/main/java/org/apache/flink/client/deployment/application/ApplicationDispatcherBootstrap.java [hunk 4]
```diff
@@ -73,35 +72,38 @@
...
 		this.configuration = checkNotNull(configuration);
-		this.recoveredJobs = checkNotNull(recoveredJobs);
+		this.recoveredJobIds = checkNotNull(recoveredJobIds);
 		this.application = checkNotNull(application);
+		this.errorHandler = checkNotNull(errorHandler);
 	}
 
 	@Override
-	public void initialize(final Dispatcher dispatcher, ScheduledExecutor scheduledExecutor) {
+	public void initialize(final DispatcherGateway dispatcher, ScheduledExecutor scheduledExecutor) {
```
Comment (tillrohrmann) [inline flink-clients/src/main/java/org/apache/flink/client/deployment/application/ApplicationDispatcherBootstrap.java:105]:
```suggestion
	public void initialize(final DispatcherGateway dispatcherGateway, ScheduledExecutor scheduledExecutor) {
```

File: flink-clients/src/test/java/org/apache/flink/client/deployment/application/ApplicationDispatcherBootstrapTest.java [hunk 5]
```diff
@@ -233,7 +234,7 @@ public void testApplicationFailsAsSoonAsOneJobFails() throws Throwable {
 
 		final CompletableFuture<Void> applicationFuture = runApplication(dispatcherBuilder, 2);
 
-		assertException(applicationFuture, JobExecutionException.class);
+		assertException(applicationFuture, ApplicationFailedException.class);
```
Comment (tillrohrmann) [inline flink-clients/src/test/java/org/apache/flink/client/deployment/application/ApplicationDispatcherBootstrapTest.java:237]: As a side comment concerning the `ApplicationDispatcherBootstrapTest`: It seems a bit strange that I couldn't find a place in the tests where we are calling `ApplicationDispatcherBootstrap.initialize()`. Looking at the interface, this looks like a very important call to do before doing any other operations on the implementations.

File: flink-clients/src/test/java/org/apache/flink/client/deployment/application/ApplicationDispatcherBootstrapTest.java [hunk 6]
```diff
@@ -454,22 +454,22 @@ public void testClusterShutdownWhenApplicationFails() throws Exception {
...
+		return createApplicationDispatcherBootstrap(noOfJobs, exception -> {});
 	}
 
 	private ApplicationDispatcherBootstrap createApplicationDispatcherBootstrap(
 			int noOfJobs,
-			Collection<JobGraph> recoveredJobGraphs) throws FlinkException {
+			FatalErrorHandler errorHandler) throws FlinkException {
 		final PackagedProgram program = getProgram(noOfJobs);
-		return new ApplicationDispatcherBootstrap(program, recoveredJobGraphs, getConfiguration());
+		return new ApplicationDispatcherBootstrap(program, Collections.emptyList(), getConfiguration(), errorHandler);
```
Comment (tillrohrmann) [inline flink-clients/src/test/java/org/apache/flink/client/deployment/application/ApplicationDispatcherBootstrapTest.java:472]: I think we are lacking a test which ensures that in case of an exceptionally completed result from `fixJobIdAndRunApplicationAsync`, which is not caused by a `FAILED` or `CANCELLED` job, we don't call `DispatcherGateway.shutDownCluster()`.

**Review by tillrohrmann — CHANGES_REQUESTED** [review]
Thanks for creating this PR @kl0u. The PR looks good to me. I had a couple of smaller comments. The one thing we should add is a test which ensures that we don't accidentally call `DispatcherGateway.shutDownCluster` in case of a fatal exception.

File: flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/Dispatcher.java [hunk 7]
```diff
@@ -208,7 +210,8 @@ public void onStart() throws Exception {
 			throw exception;
 		}
 
-		dispatcherBootstrap.initialize(this, this.getRpcService().getScheduledExecutor());
+		this.dispatcherBootstrap = this.dispatcherBootstrapFactory.apply(shutDownFuture::completeExceptionally);
```
[the remark this answers is not in the document]
Comment (kl0u) [reply flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/Dispatcher.java:218]: The reason for this is that this would bubble up to `ClusterEntrypoint` calling `System.exit(RUNTIME_FAILURE_RETURN_CODE);` and I do not know if this will kill that cluster without deleting the HA data. 

If this is this the contract, then I can do it.

File: flink-clients/src/test/java/org/apache/flink/client/deployment/application/ApplicationDispatcherBootstrapTest.java (hunk 5 above)
Comment (kl0u) [reply flink-clients/src/test/java/org/apache/flink/client/deployment/application/ApplicationDispatcherBootstrapTest.java:237]: The problem was that we were passing a `Dispatcher` which is difficult to instantiate and an overkill for such as test. This is the reason why the `runApplicationAndShutdownClusterAsync()` is annotated as `visibleForTesting` and is not private, so that all the tests can call this method.

Comment (tillrohrmann) [reply flink-clients/src/test/java/org/apache/flink/client/deployment/application/ApplicationDispatcherBootstrapTest.java:237]: The fact that this worked is actually an indicator that the `ApplicationDispatcherBootstrap` class is following a quite procedural style. Maybe with the change that it now takes a `DispatcherGateway` for which we have easy test implementations one might rethink it.

Comment (kl0u) [reply flink-clients/src/test/java/org/apache/flink/client/deployment/application/ApplicationDispatcherBootstrapTest.java]: I agree that now we can rethink it, but if I understand the comment correctly, I tend to disagree that this shows that the class follows a procedural style. The only reason that this method was exposed was for testing and it was not accessed by other classes. In that sense, internal state of the class was not "leaking" as even these methods were sealing internal state from outside tampering.

**Comment by kl0u** [comment]
Thanks for the comments @tillrohrmann ! 

I integrated them. One of your comments was to merge the `initialize()` with the constructor of the `ApplicationDispatcherBootstrap`. I am a bit torn about launching async calls in the constructor, as required by this change, but let me know what you think (it is integrated as a separate commit).

File: flink-clients/src/main/java/org/apache/flink/client/deployment/application/ApplicationDispatcherBootstrap.java [hunk 8]
```diff
@@ -124,62 +125,64 @@ public void stop() {
...
+						final ApplicationStatus applicationStatus = exception.get().getStatus();
+
+						if (applicationStatus == ApplicationStatus.CANCELED || applicationStatus == ApplicationStatus.FAILED) {
+							LOG.info("Application {}: ", applicationStatus, t);
+							return dispatcherGateway.shutDownCluster(applicationStatus);
+						}
 					}
-					return dispatcher.shutDownCluster(applicationStatus);
+
+					LOG.warn("Exiting with Application Status UNKNOWN: ", t);
```
Comment (tillrohrmann) [inline flink-clients/src/main/java/org/apache/flink/client/deployment/application/ApplicationDispatcherBootstrap.java:163]: Maybe 
```suggestion
					LOG.warn("Application failed unexpectedly: ", t);
```

File: flink-clients/src/test/java/org/apache/flink/client/deployment/application/ApplicationDispatcherBootstrapTest.java [hunk 9]
```diff
@@ -301,69 +306,132 @@ public void testApplicationIsStoppedWhenStoppingBootstrap() throws Exception {
...
+
+		// we call the error handler
+		assertException(errorHandlerFuture, ApplicationExecutionException.class);
+
+		// we return a future that is completed exceptionally
+		assertException(shutdownFuture, ApplicationExecutionException.class);
+
+		assertFalse(clusterShutdown.f0);
+		dispatcherGateway.shutDownCluster(ApplicationStatus.SUCCEEDED);
+		assertTrue(clusterShutdown.f0);
```
Comment (tillrohrmann) [inline flink-clients/src/test/java/org/apache/flink/client/deployment/application/ApplicationDispatcherBootstrapTest.java:386]: Why is this necessary? Couldn't we write the test w/o it and then directly fail if `shutDownCluster` is being called?

File: flink-clients/src/test/java/org/apache/flink/client/deployment/application/ApplicationDispatcherBootstrapTest.java [hunk 10]
```diff
@@ -373,57 +441,61 @@ public void testClusterShutdownWhenSubmissionFails() throws Exception {
...
-		// wait until the bootstrap "thinks" it's done
-		shutdownFuture.get(TIMEOUT_SECONDS, TimeUnit.SECONDS);
+		final ApplicationFailureException exception = assertException(applicationFuture, ApplicationFailureException.class);
+		assertEquals(exception.getStatus(), ApplicationStatus.UNKNOWN);
 
-		// verify that the dispatcher is actually being shut down
-		assertThat(externalShutdownFuture.get(TIMEOUT_SECONDS, TimeUnit.SECONDS), is(ApplicationStatus.FAILED));
+		assertFalse(clusterShutdown.f0);
+		dispatcherGateway.shutDownCluster(ApplicationStatus.SUCCEEDED);
+		assertTrue(clusterShutdown.f0);
```
Comment (tillrohrmann) [inline flink-clients/src/test/java/org/apache/flink/client/deployment/application/ApplicationDispatcherBootstrapTest.java:498]: Same here, why do we need this?

File: flink-clients/src/test/java/org/apache/flink/client/deployment/application/ApplicationDispatcherBootstrapTest.java (hunk 9 above)
Comment (tillrohrmann) [reply flink-clients/src/test/java/org/apache/flink/client/deployment/application/ApplicationDispatcherBootstrapTest.java:386]: Or we simply add a counter to `setClusterShutdownFunction` and assert that its value `0` at the end of the test?

**Review by tillrohrmann — APPROVED** [review]
Thanks for updating this PR @kl0u. The changes look good to me.  +1 for merging this PR.

**Comment by kl0u** [comment]
Thanks for the review @tillrohrmann ! I integrated your comments and I will merge as soon as AZP gives green.
